File: programmers/javascript/level3/단어변환.js
const begin = "hit";
const target = "cog";
const words = ["hot", "dot", "dog", "lot", "log", "cog"];

let answer = 100;
let check = [];

const diffOneLetters = (begin, words) => {
  let diffNum = 0;
  for (let i = 0; i < begin.length; i++) {
    if (begin[i] !== words[i]) diffNum += 1;
  }
  if (diffNum === 1) return 1;
  return 0;
};

const dfs = (begin, target, words, depth, check) => {
  //시작단어가 타겟단어와 같다면 답은 depth-1이다.
  if (begin == target && answer > depth) {
    answer = depth;
    console.log(check);
    return;
  }
  for (let i = 0; i < words.length; i++) {
    //방문하지 않았고 단어의 차이가 한개라면 방문한다.
    if (!check[i] && diffOneLetters(begin, words[i])) {
      depth += 1;
      check[i] = true;
      dfs(words[i], target, words, depth, check);
      check[i] = false;
      depth -= 1;
    }
  }
};

const solution = (begin, target, words) => {
  if (!words.includes(target)) {
    return 0;
  }
  for (let i = 0; i < words.length; i++) check[i] = false;
  dfs(begin, target, words, 0, check);
  return answer === 100 ? 0 : answer;
};

console.log(solution(begin, target, words));

File: python/programmers/level3/단어변환.py
'''
        dfs로 돌리면서 하나하나 나아가서 답이 target이 되는 순간 number를 리턴하자
        차이가 한개인지를 리턴하는 함수 만들기
'''


def isDiffrentOne(curWord, nextWord):
    num = 0
    for x, y in zip(curWord, nextWord):
        if x != y:
            num += 1
            if num > 1:
                return False
    if num == 1:
        return True
    return False


def solution(begin, target, words):
    global answer
    answer = 51

    def dfs(curWord, num, vis):
        global answer
        if curWord == target:
            answer = min(answer, num)
            return

        for i in range(len(words)):
            if not vis[i] and isDiffrentOne(curWord, words[i]):
                vis[i] = True
                dfs(words[i], num + 1, vis)
                vis[i] = False

    dfs(begin, 0, [False] * len(words))
    return answer if answer != 51 else 0


print(solution("hit", "cog", ['hot', 'dot', 'dog', 'lot', 'log', 'cog']))
print(solution('hit', 'cog', ['hot', 'dot', 'dog', 'lot', 'log']))
